identity/webapp/src/utility/jwt-codec.ts

```diff
@@ -0,0 +1,62 @@
+import { JwtPayload, sign, verify } from 'jsonwebtoken';
+import getConfig from 'next/config';
+
+const { serverRuntimeConfig: config } = getConfig();
+
+// we need some jwt encoding to deal with passing data to an auth0 action
+// https://auth0.com/docs/customize/actions/flows-and-triggers/login-flow/redirect-with-actions#pass-data-back-to-auth0
+
+export type RegistrationInputs = {
+  firstName: string;
+  lastName: string;
+  termsAndConditions: boolean;
+};
+
+// we first need to decode the session token we receive on redirecting from the post-login flow action
+// this token will come from request query when this handler is used in the context of the registration form i.e. req.query.session_token
+export const decodeToken = (token: string): JwtPayload | string => {
+  try {
+    const decoded = verify(token, process.env.AUTH0_ACTION_SECRET);
+    return decoded;
+  } catch (e) {
+    throw new Error('Invalid session_token in decode');
+  }
+};
+
+// we then need to add our registration form data to the token along with other details
+// this token object includes iat, iss, sub, exp and ip (from auth0 incoming token)
+// we must also include the state, which validates our ability to finish the action with /continue
+// finally we must make sure to add aud (audience) as without this the token won't be accepted by auth0
+export type RegistrationJwtPayload = Pick<
+  JwtPayload,
+  | 'iat'
+  | 'iss'
+  | 'sub'
+  | 'exp'
+  | 'aud'
+  | 'state'
+  | 'https://wellcomecollection.org/terms_agreed'
+  | 'https://wellcomecollection.org/first_name'
+  | 'https://wellcomecollection.org/last_name'
```

Comment:
Fun TS tip (not necessarily better than this, just worth knowing):

```typescript
const jwtRequiredFields = [
  'iat',
  'iss',
  'sub',
  'exp',
  'aud',
  'state',
  'https://wellcomecollection.org/terms_agreed',
  'https://wellcomecollection.org/first_name',
  'https://wellcomecollection.org/last_name'
] as const;
type RegistrationJwtPayload = Pick<JwtPayload, typeof jwtRequiredFields[number]>
```

Reply:
neat – I prefer writing commas at the ends of lines to pipes and spaces at the beginning!

Reply:
is the `as const` important? IDE not shouting at me if I take it off. I guess it must be otherwise TS would just thing `jwtRequiredFields` was an array of any strings